### PR TITLE
Monitor transaction submitted for execution

### PR DESCRIPTION
**Pull request by elgatovital**
Handles #1987

Changes proposed in this pull request:
- Persist tx data submitted to execution locally
- Get tx receipt and update tx status to pending if it was not yet indexed

@gnosis/mobile-devs

### CODE REVIEW

File: app/src/main/java/io/gnosis/safe/ui/transactions/TransactionListViewModel.kt
Comment (DmitryBespalov) [inline app/src/main/java/io/gnosis/safe/ui/transactions/TransactionListViewModel.kt:107]: I noticed that the code here depends on the internal structure of the transaction id. This should not be the case, since the id might change in the future, thus breaking this logic.

In addition, this was one of the API agreements.

Comment (DmitryBespalov) [reply app/src/main/java/io/gnosis/safe/ui/transactions/TransactionListViewModel.kt:107]: > there is no way to get safeTxHash from the transaction list response
maybe locally stored data could include the transaction's safeTxHash, backend transaction id, and ethereum tx hash?

this way, you store the data at the point of transaction submission, and then you can lookup the safeTxHash by the backend transaction id

Comment (DmitryBespalov) [reply app/src/main/java/io/gnosis/safe/ui/transactions/TransactionListViewModel.kt:107]: > api agreements

this was not documented but implied; besides, it's a good design practice

Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/ui/transactions/TransactionListViewModel.kt:107]: > this way, you store the data at the point of transaction submission, and then you can lookup the safeTxHash by the backend transaction id 

having to request tx details just to get safeTxHash would be inefficient. Better to stick with timestamps for now.

Comment (DmitryBespalov) [reply app/src/main/java/io/gnosis/safe/ui/transactions/TransactionListViewModel.kt:107]: oh no, what I mean is you store the safeTxHash from the transaction details when you execute/submit the transaction - you have that data and the transaction id beforehand. So, if you store it in advance, then you can look it up later when searching by the transaction id. No need for additional network requests

File: app/src/main/java/io/gnosis/safe/ui/transactions/details/TransactionDetailsViewModel.kt
Comment (rmeissner) [inline app/src/main/java/io/gnosis/safe/ui/transactions/details/TransactionDetailsViewModel.kt:280]: Should this code be removed before this PR can be merged?

**Review by rmeissner — CHANGES_REQUESTED** [review]
Please add tests for the new repository

Edit: Also a couple view models have new logic which should also be tested

**Comment by elgatovital** [comment]
> Please add tests for the new repository
> 
> Edit: Also a couple view models have new logic which should also be tested

yes, already have task for it https://github.com/safe-global/safe-android/issues/1963

**Comment by rmeissner** [comment]
We should not merge new logic without tests. It makes little sense to do the tests in a separate PR :thinking:

**Review by biafra23 — APPROVED** [review]
LGTM 🚀

**Review by DmitryBespalov — APPROVED** [review]

File: app/src/test/java/io/gnosis/safe/ui/transactions/details/TransactionDetailsViewModelTest.kt
Comment (rmeissner) [inline app/src/test/java/io/gnosis/safe/ui/transactions/details/TransactionDetailsViewModelTest.kt:124]: I would expect in the test that we check that this method was called. This is a crucial step and easy to verify via the  mocks

File: app/src/test/java/io/gnosis/safe/ui/transactions/TransactionListViewModelTest.kt
Comment (rmeissner) [inline app/src/test/java/io/gnosis/safe/ui/transactions/TransactionListViewModelTest.kt:2]: We made quite extensive changes to this view model, why is this not reflected in the tests? (i.e. we call updateLocalTxLatest and added new mappings)

File: app/src/test/java/io/gnosis/safe/ui/transactions/details/TransactionDetailsViewModelTest.kt
Comment (rmeissner) [inline app/src/test/java/io/gnosis/safe/ui/transactions/details/TransactionDetailsViewModelTest.kt:199]: Same here, `delete` being called is critical we should verify this.

**Review by rmeissner — CHANGES_REQUESTED** [review]

File: app/src/test/java/io/gnosis/safe/ui/transactions/TransactionListViewModelTest.kt
Comment (elgatovital) [reply app/src/test/java/io/gnosis/safe/ui/transactions/TransactionListViewModelTest.kt:2]: added tests

File: app/src/test/java/io/gnosis/safe/ui/transactions/details/TransactionDetailsViewModelTest.kt
Comment (rmeissner) [inline app/src/test/java/io/gnosis/safe/ui/transactions/details/TransactionDetailsViewModelTest.kt:175]: To be honest this is not a good verification. We should know exactly what values it is being called with. This is not something blocking the PR anymore, but this is not the test quality I would expect for critical flows.

File: app/src/test/java/io/gnosis/safe/ui/transactions/details/TransactionDetailsViewModelTest.kt
Comment (rmeissner) [inline app/src/test/java/io/gnosis/safe/ui/transactions/details/TransactionDetailsViewModelTest.kt:329]: Same comment here as above. We should verify that exactly the delete is called that we expect. The most complete test would be to check that the method with any parameter was called exactly once and that the parameter was the expected transaction.

**Pull request merged**
